### PR TITLE
UCT/IB/MLX5/DV: fixed memic iova in ksm registration

### PR DESCRIPTION
**Pull request by roiedanino**
## What
In ksm registration: setting iova to `(uint64_t)memh->address` in case of memic.
In perftest: check allocators are not NULL before initializing them

## Why ?
MEMIC-relate bugfixes - 
1. Fix in ksm registration is necessary for rdma memory to work
2. Fix in perftest avoids a segfault when using multiple threads with RDMA memory

### CODE REVIEW

File: src/tools/perf/lib/libperf_thread.c [hunk 1]
```diff
@@ -148,15 +148,22 @@ ucs_status_t ucx_perf_thread_spawn(ucx_perf_context_t *perf,
 
 ucs_status_t ucx_perf_allocators_init_thread(ucx_perf_context_t *perf)
 {
+    ucs_status_t status;
     /* New threads need explicit device association */
-    ucs_status_t status = perf->send_allocator->init(perf);
-    if (status != UCS_OK) {
-        return status;
+    if (perf->send_allocator != NULL) {
```
[the remark this answers is not in the document]
Comment (tvegas1) [reply src/tools/perf/lib/libperf_thread.c:153]: if it has to be null we could use template always returning `UCS_OK` instead.

[the remark this answers is not in the document]
Comment (roiedanino) [reply src/tools/perf/lib/libperf_thread.c:153]: @yosefe When using RDMA memory:
> Fix in perftest avoids a segfault when using multiple threads with RDMA memory

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 2]
```diff
@@ -237,20 +237,22 @@ static ucs_status_t uct_ib_mlx5_devx_reg_ksm_data_addr(
...
-    uint64_t ksm_iova;
+    uint64_t ksm_iova, ksm_iova_base;
     size_t ksm_length;
     int list_size;
 
     /* FW requires indirect atomic MR address and length to be aligned
      * to max supported atomic argument size */
     ksm_address = ucs_align_down_pow2((uint64_t)address, UCT_IB_MD_MAX_MR_SIZE);
-    ksm_iova    = ksm_address + iova_offset;
+    ksm_iova_base = memh->dm != NULL ? (uint64_t)memh->address : ksm_address;
```
[the remark this answers is not in the document]
Comment (roiedanino) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:254]: When using RDMA memory (dm != NULL), we want to map the reserved address range (memh->address) to 0 which is the actual RDMA base address for it.

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 3]
```diff
@@ -237,20 +237,22 @@ static ucs_status_t uct_ib_mlx5_devx_reg_ksm_data_addr(
  * that were registered in single-threaded mode.
  */
 static ucs_status_t uct_ib_mlx5_devx_reg_ksm_data_contig(
```
[the remark this answers is not in the document]
Comment (roiedanino) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:239]: Will change it as well, thanks

File: src/tools/perf/lib/libperf_thread.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/tools/perf/lib/libperf_thread.c:153]: > @yosefe When using RDMA memory:
> 
> > Fix in perftest avoids a segfault when using multiple threads with RDMA memory

Is this a PR/commit? can you pls point to it?

[the remark this answers is not in the document]
Comment (roiedanino) [reply src/tools/perf/lib/libperf_thread.c:153]: 2 commits caused different issues here:
6fbfdea4f89d7cc09e7ba6ed1971329c68ba00f7 - assumed `iova` is always `= address + offset` while when using memic the base address should be 0. 

fc85a133f3f84feda80d33ef6a5ce1bb52916e1c - creating all self_eps with the same worker address (issue was identified only when using multiple threads obviously).

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 4]
```diff
@@ -432,16 +446,17 @@ uct_ib_mlx5_devx_reg_ksm_data(uct_ib_mlx5_md_t *md,
...
     if (memh->super.flags & UCT_IB_MEM_MULTITHREADED) {
         return uct_ib_mlx5_devx_reg_ksm_data_mt(md, address,
                                                 (uint64_t)address + iova_offset,
                                                 atomic, mkey_index, reason,
                                                 mr->ksm_data, mr_p, mkey);
     } else {
-        return uct_ib_mlx5_devx_reg_ksm_data_contig(md, address, iova_offset,
-                                                    atomic, mkey_index, reason,
-                                                    mr, mr_p, mkey);
+        return uct_ib_mlx5_devx_reg_ksm_data_contig(md, address, iova, atomic,
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:457]: shouldn't we pass iova to uct_ib_mlx5_devx_reg_ksm_data_mt (line 452) as well? instead of `(uint64_t)address + iova_offset` ?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 5]
```diff
@@ -276,6 +274,22 @@ uct_ib_mlx5_devx_memh_base_address(const uct_ib_mlx5_devx_mem_t *memh)
...
+
+#if HAVE_IBV_DM
+    if (memh->dm != NULL) {
+        /* Device memory memory key is zero based */
+        return (uint64_t)memh->address + iova_offset;
+    }
+#endif
+
+    address = (uint64_t)uct_ib_mlx5_devx_memh_base_address(memh);
+    return ucs_align_down_pow2(address, UCT_IB_MD_MAX_MR_SIZE) + iova_offset;
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:290]: uct_ib_mlx5_devx_reg_ksm_data_contig should adjust the iova it receives according to alignment - instead of doing it here.

File: src/tools/perf/lib/libperf_thread.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/tools/perf/lib/libperf_thread.c:153]: > 2 commits caused different issues here: [6fbfdea](https://github.com/openucx/ucx/commit/6fbfdea4f89d7cc09e7ba6ed1971329c68ba00f7) - assumed `iova` is always `= address + offset` while when using memic the base address should be 0.
> 
> [fc85a13](https://github.com/openucx/ucx/commit/fc85a133f3f84feda80d33ef6a5ce1bb52916e1c) - creating all self_eps with the same worker address (issue was identified only when using multiple threads obviously).

1. According to https://github.com/openucx/ucx/blob/8668850ca49311024949c793b18ea5ec24b96987/src/tools/perf/lib/libperf.c#L2074-L2079, send_allocator and recv_allocator cannot be NULL during the test. So why need to check for NULL here?

2. How this segfault was not caught in CI? Maybe we need to add multi-threaded perftest?

File: src/tools/perf/lib/libperf.c [hunk 6]
```diff
@@ -1520,26 +1520,32 @@ static ucs_status_t ucp_perf_test_flush_workers(ucx_perf_context_t *perf)
...
+        if (status != UCS_OK) {
+            if (i > 0) {
+                goto err_destroy_eps;
+            }
+            return status;
+        }
+
+        ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+        ep_params.address    = worker_attr.address;
+
```
Comment (yosefe) [inline src/tools/perf/lib/libperf.c:1548]: remove space line

File: src/tools/perf/lib/libperf.c [hunk 7]
```diff
@@ -1520,26 +1520,32 @@ static ucs_status_t ucp_perf_test_flush_workers(ucx_perf_context_t *perf)
...
-
-    ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
-    ep_params.address    = worker_attr.address;
-
     for (i = 0; i < perf->params.thread_count; ++i) {
         thread_perf = &perf->ucp.tctx[i].perf;
+        worker      = thread_perf->ucp.worker;
+        /* coverity[use_after_free] */
+        status      = UCX_PERF_VERBOSE(error, &perf->params, ucp_worker_query,
+                                       worker, &worker_attr);
```
Comment (yosefe) [inline src/tools/perf/lib/libperf.c:1538]: So the problem here was that ep from all threads were connected to the "main" worker and not the "thread" worker? How is this fix related to MEMIC?

File: src/tools/perf/lib/libperf.c [hunk 8]
```diff
@@ -1520,26 +1520,32 @@ static ucs_status_t ucp_perf_test_flush_workers(ucx_perf_context_t *perf)
...
+        if (status != UCS_OK) {
+            if (i > 0) {
+                goto err_destroy_eps;
```
Comment (yosefe) [inline src/tools/perf/lib/libperf.c:1541]: maybe we can always `goto err_destroy_eps` in case of error?

File: src/tools/perf/lib/libperf.c [hunk 9]
```diff
@@ -1557,6 +1563,8 @@ static ucs_status_t ucp_perf_test_setup_self_endpoints(ucx_perf_context_t *perf)
         if (status != UCS_OK) {
             goto err_destroy_eps;
         }
+
+        ucp_worker_release_address(worker, worker_attr.address);
```
Comment (yosefe) [inline src/tools/perf/lib/libperf.c:1567]: should we set address to NULL after we release it, to avoid double-free in line 1580?
(Maybe Coverity warning was justified?)

Comment (roiedanino) [reply src/tools/perf/lib/libperf.c:1567]: Sounds good

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 4 above)
Comment (roiedanino) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:457]: I think not, because data_contig is aligning down the base address while it was not the logic in `uct_ib_mlx5_devx_reg_ksm_data_mt`

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 5 above)
Comment (roiedanino) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:290]: So we will have to pass `iova_offset` separately to ksm_data_contig

File: src/tools/perf/lib/libperf.c (hunk 7 above)
Comment (yosefe) [reply src/tools/perf/lib/libperf.c:1538]: maybe it has to be a separate fix in another PR?
does it have some dependency with MEMIC fix?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 5 above)
Comment (yosefe) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:290]: instead, maybe we could add the difference between address and ksm_address (i.e how much alignment has moved the address) to iova argument?

**Comment by roiedanino** [comment]
@yosefe 
https://github.com/openucx/ucx/pull/9743#discussion_r1530249046

This logic explains it:
```c
ucs_status_t ucx_perf_allocators_init(ucx_perf_context_t *perf,
                                      const ucx_perf_params_t *params)
{
    if ((params->send_mem_type == UCS_MEMORY_TYPE_RDMA) ||
        (params->recv_mem_type == UCS_MEMORY_TYPE_RDMA)) {
        return UCS_OK;
    }
```
We need the allocators initialization also in UCP for the CUDA / ROCM context. the multithreading logic is shared with the UCT perftests which are using the allocators in runtime, I think we can't avoid that null check.

File: src/tools/perf/lib/libperf.c (hunk 7 above)
Comment (roiedanino) [reply src/tools/perf/lib/libperf.c:1538]: We can split it but we won't be able to test memic with multithreading using just one of those fixes, what about separating this PR into 2 different commits when squashing?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 10]
```diff
@@ -276,6 +274,22 @@ uct_ib_mlx5_devx_memh_base_address(const uct_ib_mlx5_devx_mem_t *memh)
...
 
+static uint64_t uct_ib_mlx5_devx_memh_iova(const uct_ib_mlx5_devx_mem_t *memh,
+                                           uint32_t iova_offset)
+{
+    uint64_t address;
+
+#if HAVE_IBV_DM
+    if (memh->dm != NULL) {
+        /* Device memory memory key is zero based */
+        return (uint64_t)memh->address + iova_offset;
```
Comment (ivankochin) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:285]: Should we align to `memh->address` to `UCT_IB_MD_MAX_MR_SIZE`  in that case?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 4 above)
Comment (ivankochin) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:457]: After one of the recent patches it does (but it aligns to `chunk_size` not to `UCT_IB_MD_MAX_MR_SIZE`):

https://github.com/openucx/ucx/blob/master/src/uct/ib/mlx5/dv/ib_mlx5dv_md.c#L162

File: src/tools/perf/lib/libperf.c (hunk 7 above)
Comment (roiedanino) [reply src/tools/perf/lib/libperf.c:1537]: Will make this PR the memic fix and open a new PR for the perftest

Comment (roiedanino) [reply src/tools/perf/lib/libperf.c:1537]: https://github.com/openucx/ucx/pull/9766

**Review by yosefe — COMMENTED** [review]
can we add a gtest to expose the bug?

**Comment by roiedanino** [comment]
> can we add a gtest to expose the bug?

We can create the test but it will only fail CI when the new FW will be used in CI

**Comment by yosefe** [comment]
> > can we add a gtest to expose the bug?
> 
> We can create the test but it will only fail CI when the new FW will be used in CI

Let's add it (van test on rock cluster in meantime)

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 10 above)
Comment (roiedanino) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:285]: No

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 11]
```diff
@@ -241,20 +241,18 @@ static ucs_status_t uct_ib_mlx5_devx_reg_ksm_data_addr(
...
     size_t mr_length = mr->super.ib->length;
     uint64_t ksm_address;
-    uint64_t ksm_iova;
     size_t ksm_length;
     int list_size;
 
     /* FW requires indirect atomic MR address and length to be aligned
      * to max supported atomic argument size */
     ksm_address = ucs_align_down_pow2((uint64_t)address, UCT_IB_MD_MAX_MR_SIZE);
-    ksm_iova    = ksm_address + iova_offset;
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:257]: `ksm_iova = ksm_address - address + iova;`

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 4 above)
Comment (roiedanino) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:446]: Recent update should be good for both cases

**Review by yosefe — APPROVED** [review]

**Comment by ivankochin** [comment]
> > > can we add a gtest to expose the bug?
> > 
> > 
> > We can create the test but it will only fail CI when the new FW will be used in CI
> 
> Let's add it (can test on rock cluster in meantime)

What about adding gtest for that fix? If I understand the quoted discussion correctly, it should be added, isn't it?

**Comment by yosefe** [comment]
> > > > can we add a gtest to expose the bug?
> > > 
> > > 
> > > We can create the test but it will only fail CI when the new FW will be used in CI
> > 
> > 
> > Let's add it (can test on rock cluster in meantime)
> 
> What about adding gtest for that fix? If I understand the quoted discussion correctly, it should be added, isn't it?

We decided to add tests in another PR since this fix is needed for the release
